File: scripts/check-fonts.py
from pathlib import Path

from fontTools.ttLib import TTFont


ROOT = Path(__file__).resolve().parent.parent
SAMPLE = set("温油站主题正文创作骰子列表消息注册登录")


def codepoints(font: TTFont) -> set[int]:
    result: set[int] = set()
    for table in font["cmap"].tables:
        result.update(table.cmap)
    return result


fonts = {
    "Noto Sans SC": ROOT / "packages/flutter/lib/fonts/NotoSansSC-Variable.ttf",
    "LXGW WenKai": ROOT / "packages/flutter/lib/fonts/LXGWWenKaiLite-Medium.ttf",
    "Nunito": ROOT / "packages/flutter/lib/fonts/Nunito-Variable.ttf",
    "LXGW WenKai Web": ROOT / "web/fonts/LXGWWenKaiLite-Medium.woff2",
}

for name, path in fonts.items():
    font = TTFont(path)
    if name != "Nunito":
        missing = sorted(
            character for character in SAMPLE if ord(character) not in codepoints(font)
        )
        if missing:
            raise SystemExit(f"{name} 缺少代表性中文字形：{''.join(missing)}")
    if name in {"Noto Sans SC", "Nunito"} and "fvar" not in font:
        raise SystemExit(f"{name} 必须保留可变字体轴")

print("Font assets contain expected Chinese glyphs and variable axes")

File: scripts/check-fonts.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";
import * as fontkit from "fontkit";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const sample = [..."温油站主题正文创作骰子列表消息注册登录"];
const fonts = new Map([
  ["Noto Sans SC", "packages/flutter/lib/fonts/NotoSansSC-Variable.ttf"],
  ["LXGW WenKai", "packages/flutter/lib/fonts/LXGWWenKaiLite-Medium.ttf"],
  ["Nunito", "packages/flutter/lib/fonts/Nunito-Variable.ttf"],
  ["LXGW WenKai Web", "web/fonts/LXGWWenKaiLite-Medium.woff2"],
]);

for (const [name, relativePath] of fonts) {
  const font = fontkit.openSync(path.join(root, relativePath));
  if (name !== "Nunito") {
    const missing = sample.filter((character) => !font.hasGlyphForCodePoint(character.codePointAt(0)));
    if (missing.length > 0) throw new Error(`${name} 缺少代表性中文字形：${missing.join("")}`);
  }
  if (["Noto Sans SC", "Nunito"].includes(name) && Object.keys(font.variationAxes ?? {}).length === 0) {
    throw new Error(`${name} 必须保留可变字体轴`);
  }
}

console.log("Font assets contain expected Chinese glyphs and variable axes");
